### PR TITLE
feat: support aggregate function retention

### LINKED ISSUE #4773: Feature: support aggregate function retention
**Summary**

Description for this feature.

https://clickhouse.com/docs/en/sql-reference/aggregate-functions/parametric-functions/#retention

### PR DESCRIPTION
**Pull request by fkuner**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

support aggregate function retention

## Changelog

- New Feature
- Documentation

## Related Issues

Fixes #4773

### CODE REVIEW

File: common/functions/src/aggregates/aggregate_retention.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        write!(f, "{}", self.display_name)
+    }
+}
+
+impl AggregateRetentionFunction {
+    pub fn try_create(
+        display_name: &str,
+        arguments: Vec<DataField>,
+    ) -> Result<AggregateFunctionRef> {
+        Ok(Arc::new(Self {
```
Comment (sundy-li) [inline common/functions/src/aggregates/aggregate_retention.rs:188]: Validate the `arguments`

File: common/functions/src/aggregates/aggregate_retention.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        _validity: Option<&common_arrow::arrow::bitmap::Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+
+        for i in 0..input_rows {
+            for j in 0..self.events_size {
+                let column: &BooleanColumn = Series::check_get(&columns[j as usize])?;
```
Comment (sundy-li) [inline common/functions/src/aggregates/aggregate_retention.rs:90]: Make this `column` be prepared before the loop!

Comment (fkuner) [reply common/functions/src/aggregates/aggregate_retention.rs:90]: ok!

File: common/functions/src/aggregates/aggregate_retention.rs (hunk 1 above)
Comment (fkuner) [reply common/functions/src/aggregates/aggregate_retention.rs:188]: ok!

File: docs/doc/30-reference/20-functions/50-aggregate-functions/aggregate-retention.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+| cond  | An expression that returns a Boolean result |
+
+## Return Type
+
+The array of 1 or 0.
+
+## Examples
+
+```
+CREATE TABLE retention_test(date Date, uid Int32) ENGINE = Memory;
```
Comment (sundy-li) [inline docs/doc/30-reference/20-functions/50-aggregate-functions/aggregate-retention.md:32]: Let's use the sql style name.

https://github.com/datafuselabs/databend/blob/d3c0139e18cc98b97a82735df28f6693bbc85f70/docs/doc/30-reference/10-data-types/index.md

Comment (fkuner) [reply docs/doc/30-reference/20-functions/50-aggregate-functions/aggregate-retention.md:32]: ok

**Pull request closed without merging**